### PR TITLE
[SHELL32][SDK] Implement SHOpenEffectiveToken and SHGetUserSessionId

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

Implementing missing features...
JIRA issue: N/A

## Proposed changes

- Add `OpenEffectiveToken` helper function.
- Implement `SHOpenEffectiveToken` and `SHGetUserSessionId` functions.
- Add them to `<undocshell.h>`.

## TODO

- [x] Do build.

### CODE REVIEW

File: dll/win32/shell32/utils.cpp [hunk 1]
```diff
@@ -9,6 +9,56 @@
...
+EXTERN_C BOOL WINAPI SHOpenEffectiveToken(_Out_ LPHANDLE phToken)
+{
+    TRACE("%p\n", phToken);
+    return OpenEffectiveToken((TOKEN_QUERY | TOKEN_ADJUST_PRIVILEGES), phToken);
+}
+
+/*************************************************************************
+ *                SHGetUserSessionId (SHELL32.248)
+ */
+EXTERN_C DWORD WINAPI SHGetUserSessionId(_In_ HANDLE hToken)
```
Comment (binarymaster) [inline dll/win32/shell32/utils.cpp:43]:
```suggestion
EXTERN_C DWORD WINAPI SHGetUserSessionId(_In_opt_ HANDLE hToken)
```

File: dll/win32/shell32/utils.cpp [hunk 2]
```diff
@@ -9,6 +9,56 @@
...
+    return ret;
+}
+
+/*************************************************************************
+ *                SHOpenEffectiveToken (SHELL32.235)
+ */
+EXTERN_C BOOL WINAPI SHOpenEffectiveToken(_Out_ LPHANDLE phToken)
+{
+    TRACE("%p\n", phToken);
+    return OpenEffectiveToken((TOKEN_QUERY | TOKEN_ADJUST_PRIVILEGES), phToken);
```
Comment (binarymaster) [inline dll/win32/shell32/utils.cpp:37]:
```suggestion
    return OpenEffectiveToken(TOKEN_QUERY | TOKEN_ADJUST_PRIVILEGES, phToken);
```

File: dll/win32/shell32/utils.cpp [hunk 3]
```diff
@@ -9,6 +9,56 @@
 
 WINE_DEFAULT_DEBUG_CHANNEL(shell);
 
+static BOOL OpenEffectiveToken(DWORD DesiredAccess, HANDLE *phToken)
+{
+    BOOL ret;
+
+    if (IsBadWritePtr(phToken, sizeof(*phToken)))
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:16]: This is a baroque function, not sure it has to be used there. OpenThreadToken/OpenProcessToken should already do validation checks and return a suitable error there.

File: dll/win32/shell32/utils.cpp [hunk 4]
```diff
@@ -9,6 +9,56 @@
...
+    DWORD dwSessionId = 0, dwLength;
+    BOOL bOpenToken = FALSE;
+
+    TRACE("%p\n", hToken);
+
+    if (!hToken)
+        bOpenToken = SHOpenEffectiveToken(&hToken);
+
+    if (hToken)
+        GetTokenInformation(hToken, TokenSessionId, &dwSessionId, sizeof(dwSessionId), &dwLength);
```
Comment (HBelusca) [inline dll/win32/shell32/utils.cpp:54]: There is no return error check being done there?

File: dll/win32/shell32/utils.cpp (hunk 3 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp]: Fixed.

File: dll/win32/shell32/utils.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp]: Checked.

File: dll/win32/shell32/utils.cpp [hunk 5]
```diff
@@ -9,6 +9,59 @@
...
+    TRACE("%p\n", phToken);
+    return OpenEffectiveToken(TOKEN_QUERY | TOKEN_ADJUST_PRIVILEGES, phToken);
+}
+
+/*************************************************************************
+ *                SHGetUserSessionId (SHELL32.248)
+ */
+EXTERN_C DWORD WINAPI SHGetUserSessionId(_In_opt_ HANDLE hToken)
+{
+    DWORD dwSessionId = 0, dwLength;
```
Comment (binarymaster) [inline dll/win32/shell32/utils.cpp:45]:
```suggestion
    DWORD dwSessionId, dwLength;
```

File: sdk/include/reactos/undocshell.h [hunk 6]
```diff
@@ -684,6 +684,9 @@ LargeIntegerToString(
     _In_opt_ const NUMBERFMTW *pNumberFormat,
     _In_ DWORD dwNumberFlags);
 
+BOOL WINAPI SHOpenEffectiveToken(_Out_ LPHANDLE phToken);
+DWORD WINAPI SHGetUserSessionId(_In_ HANDLE hToken);
```
Comment (binarymaster) [inline sdk/include/reactos/undocshell.h:688]:
```suggestion
DWORD WINAPI SHGetUserSessionId(_In_opt_ HANDLE hToken);
```

File: dll/win32/shell32/utils.cpp (hunk 5 above)
Comment (katahiromz) [reply dll/win32/shell32/utils.cpp]: Fixed.

**Pull request merged**